modules/core/arrow-annotations-processor/src/main/java/arrow/renzu/RenzuGenerator.kt

```diff
@@ -0,0 +1,181 @@
+    if (other !is TypeClass) false else simpleName == other.simpleName
+
+  override fun hashCode(): Int = simpleName.hashCode()
+
+  /**
+   * Returns the typeclasses that are a direct parent of this one, for composing the typeclass
+   * relationship on the tree.
+   */
+  fun parentTypeClasses(current: ClassOrPackageDataWrapper.Class): List<TypeClass> {
+    val typeTable = TypeTable(current.classProto.typeTable)
+    val interfaces = current.classProto.supertypes(typeTable).map {
+      it.extractFullName(current)
+    }.filter {
+      it != "`kotlin`.`Any`"
+    }
+    val parentInterfaces: List<ClassOrPackageDataWrapper.Class> = listOf()
+    return when {
+      interfaces.isEmpty() -> parentInterfaces.map {
+        TypeClass(processor, it.simpleName, it)
+      }
+      else -> {
+        interfaces.flatMap { i ->
+          try {
+            val className = i.removeBackticks().substringBefore("<")
+            val typeClassElement = processor.elementUtils.getTypeElement(className)
+            val parentInterface = processor.getClassOrPackageDataWrapper(typeClassElement) as ClassOrPackageDataWrapper.Class
+            if (i.removeBackticks().contains("typeclass")) {
+              parentInterfaces + parentInterface
+            } else
+              parentInterfaces
+          } catch (_: Throwable) {
+            emptyList<ClassOrPackageDataWrapper.Class>()
+          }
+        }.map {
+          TypeClass(processor, it.simpleName, it)
+        }
+      }
+    }
+  }
+}
+
+typealias ParentTypeClass = TypeClass
+typealias Instances = Set<Instance>
+
+class RenzuGenerator(
+  private val processor: RenzuProcessor,
+  annotatedList: List<AnnotatedInstance>,
+  val isolateForTests: Boolean = false) {
+
+  private val typeclassTree: MutableMap<TypeClass, Tuple2<Instances, Set<ParentTypeClass>>> =
+    normalizeTypeclassTree(annotatedList.map { Instance(it) })
+
+  private fun normalizeTypeclassTree(instances: List<Instance>)
+    : MutableMap<TypeClass, Tuple2<Instances, Set<ParentTypeClass>>> =
+    instances.fold(mutableMapOf()) { acc, instance ->
+      instance.implementedTypeclasses().forEach { tc ->
+        val typeclass = TypeClass(processor, tc.simpleName, tc)
+        val parentTypeClasses = typeclass.parentTypeClasses(tc)
+
+        acc.computeIfPresent(typeclass,
+          { _, value ->
+            Tuple2(
+              value._1 + setOf(instance),
+              value._2 + parentTypeClasses)
+          })
+        acc.putIfAbsent(typeclass,
+          Tuple2(setOf(instance), parentTypeClasses.toSet()))
+      }
+      acc
+    }
+
+  fun generate() {
+    val generatedDir = if (isolateForTests)
+      File("./infographic")
+    else
+      File("${recurseFilesUpwards("arrow").absolutePath}/infographic", "")
+        .also { it.mkdirs() }
+
+    val file = File(generatedDir, "arrow-infographic.txt")
+    if (!file.exists()) {
+      val globalStyles =
+        listOf("#font: Menlo") +
```

Comment:
This can be probably  expressed as a string template:
```kotlin
val globalStyles = 
  """
     |#font: Menlo
     |...
     """.trimMarging()
```